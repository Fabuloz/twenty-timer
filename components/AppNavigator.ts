import { createAppContainer, createStackNavigator } from 'react-navigation';
import {
  AlertsScreen,
  ColorsScreen,
  HomeScreen,
  SettingsScreen
} from '../screens';

// Maps links to colors and alerts settings screens
const SettingsNavigator = createStackNavigator(
  {
    Settings: {
      screen: SettingsScreen,
      navigationOptions: {
        gesturesEnabled: false
      }
    },
    Colors: {
      screen: ColorsScreen,
      navigationOptions: {
        gesturesEnabled: false
      }
    },
    Alerts: {
      screen: AlertsScreen,
      navigationOptions: {
        gesturesEnabled: false
      }
    }
  },
  {
    headerMode: 'none'
  }
);

// Maps link between home and settings screen
const AppNavigator = createStackNavigator(
  {
    Home: {
      screen: HomeScreen,
      navigationOptions: {
        gesturesEnabled: false
      }
    },
    Settings: SettingsNavigator
  },
  {
    headerMode: 'none',
    initialRouteName: 'Home',
    mode: 'modal'
  }
);

export default createAppContainer(AppNavigator);
