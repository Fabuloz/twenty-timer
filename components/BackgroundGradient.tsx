import React, { memo, ReactElement } from 'react';
import { LinearGradient } from 'expo-linear-gradient';
import AppStyles from '../styles/AppStyles';

/**
 * Linear gradient background
 */
const BackgroundGradient = (): ReactElement => {
  return (
    <LinearGradient
      colors={['transparent', 'rgba(0,0,0,.65)']}
      style={AppStyles.gradient}
    />
  );
};

export default memo(BackgroundGradient);
