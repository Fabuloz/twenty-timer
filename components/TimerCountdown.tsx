import React, { ReactElement, useEffect, useRef, useState } from 'react';
import { Text, TextProps } from 'react-native';

/**
 * TimerCountdown component prop types
 */
interface TimerCountdownProps extends TextProps {
  initialSeconds: number;
  onTimeElapsed: VoidFunction;
}

/**
 * Timer countdown component
 * @param initialSeconds - Initial countdown timer start time
 * @param onTimeElapsed - Function to call upon countdown completion
 */
const TimerCountdown = ({
  initialSeconds,
  onTimeElapsed,
  ...rest
}: TimerCountdownProps): ReactElement => {
  // Current seconds on timer
  const secondsRef = useRef<number>();
  const [seconds, setSeconds] = useState<number>(initialSeconds);

  // Interval id
  const intervalRef = useRef<number>();

  // Tick one second
  const tick = (): void => {
    if (secondsRef.current === 0) {
      onTimeElapsed();
      clearInterval(intervalRef.current);
      secondsRef.current = undefined;
    } else if (secondsRef.current !== undefined) {
      setSeconds(--secondsRef.current);
    }
  };

  // Format remaining seconds like `minutes`:`seconds`
  const getFormattedTime = (secs: number): string => {
    const minutes = Math.floor((secs % (60 * 60)) / 60);
    const seconds = Math.ceil((secs % (60 * 60)) % 60);

    const min = minutes < 10 ? `0${minutes}` : minutes;
    const sec = seconds < 10 ? `0${seconds}` : seconds;

    return `${min}:${sec}`;
  };

  useEffect((): VoidFunction => {
    if (secondsRef.current === undefined) {
      setSeconds((secondsRef.current = initialSeconds));
    }

    const id = setInterval(tick, 1000);
    intervalRef.current = id;

    return (): void => {
      clearInterval(intervalRef.current);
    };
  }, [initialSeconds, onTimeElapsed]);

  return <Text {...rest}>{getFormattedTime(seconds)}</Text>;
};

export default TimerCountdown;
