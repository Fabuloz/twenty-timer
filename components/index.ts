// components barrel file
export { default as AppNavigator } from './AppNavigator';
export { default as BackgroundGradient } from './BackgroundGradient';
export { default as EntypoNavigator } from './EntypoNavigator';
export { default as TimerCountdown } from './TimerCountdown';
