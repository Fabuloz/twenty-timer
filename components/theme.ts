// App color theme
const theme: { [key: string]: string } = {
  brown: '#4b250c',
  green: '#4b450c',
  white: '#fffbf7'
};

export default theme;
