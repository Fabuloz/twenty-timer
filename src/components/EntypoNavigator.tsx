import React, { memo, ReactElement } from 'react';
import { TouchableOpacity, View, TouchableOpacityProps } from 'react-native';
import { Entypo } from '@expo/vector-icons';
import { scale } from 'react-native-size-matters';
import AppStyles from '../styles/AppStyles';

// EntypoNavigator component prop types
interface EntypoNavigatorProps extends TouchableOpacityProps {
  name: string;
}

/**
 * Navigation icon
 * @param name - The name of the icon to render.
 */
const EntypoNavigator = ({
  name,
  ...rest
}: EntypoNavigatorProps): ReactElement => {
  return (
    <TouchableOpacity
      testID="entypo-navigator"
      style={AppStyles.navigator}
      {...rest}
    >
      <View>
        <Entypo name={name} size={scale(50)} style={AppStyles.icon} />
      </View>
    </TouchableOpacity>
  );
};

export default memo(EntypoNavigator);
