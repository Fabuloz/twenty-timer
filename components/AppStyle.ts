import { StyleSheet } from 'react-native';
import theme from './theme';

// Global application style
const AppStyle = StyleSheet.create({
  gradient: {
    height: '100%',
    left: 0,
    position: 'absolute',
    right: 0,
    top: 0
  },
  icon: {
    color: theme.white
  },
  navigator: {
    bottom: '5%',
    position: 'absolute',
    right: '5%'
  },
  screen: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center'
  }
});

export default AppStyle;
