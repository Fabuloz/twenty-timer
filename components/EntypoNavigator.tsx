import React, { memo, ReactElement } from 'react';
import { TouchableOpacity, View } from 'react-native';
import { Entypo } from '@expo/vector-icons';
import { scale } from 'react-native-size-matters';
import AppStyle from '../styles/AppStyles';

// EntypoNavigator component prop types
interface EntypoNavigatorProps {
  name: string;
  onPress: () => boolean;
}

/**
 * Navigation icon
 * @param name - The name of the icon to render.
 */
const EntypoNavigator = ({
  name,
  onPress
}: EntypoNavigatorProps): ReactElement => {
  return (
    <TouchableOpacity onPress={onPress} style={AppStyle.navigator}>
      <View>
        <Entypo name={name} size={scale(50)} style={AppStyle.icon} />
      </View>
    </TouchableOpacity>
  );
};

export default memo(EntypoNavigator);
