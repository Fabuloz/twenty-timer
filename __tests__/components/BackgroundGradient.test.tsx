import React from 'react';
import { render, toJSON } from '@testing-library/react-native';
import { BackgroundGradient } from '../../components';

test('renders and matches snapshot', (): void => {
  const { container, getByTestId } = render(<BackgroundGradient />);

  expect(getByTestId('background-gradient')).toBeTruthy();
  expect(toJSON(container)).toMatchSnapshot();
});
