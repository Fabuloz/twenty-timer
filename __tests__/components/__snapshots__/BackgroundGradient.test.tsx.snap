// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and matches snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      Object {
        "colors": Array [
          0,
          2785017856,
        ],
        "endPoint": undefined,
        "locations": undefined,
        "startPoint": undefined,
      }
    }
    style={
      Object {
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="background-gradient"
  />
</View>
`;
